ttnn/cpp/ttnn/types.hpp

```diff
@@ -59,76 +59,70 @@ struct CoreGrid {
-                this->allocate();
+                // Private constructor used by factory method. Each buffer is given a unique id.
+                this->buffer_instance_id = Buffer::buf_id++;
             }
-        ~Buffer() {
-            this->deallocate();
-        }
-    private:
-        uint32_t buffer_id = 0;
+
         void allocate() {
             TT_ASSERT(this->device());
-            this->device()->push_work([this] () mutable {
-                bool bottom_up = this->buffer_type() == BufferType::DRAM;
-                tt::tt_metal::detail::AllocateBuffer(this, bottom_up);
+            this->device()->push_work([shared_buffer_ptr = shared_from_this()] () mutable {
+                bool bottom_up = shared_buffer_ptr->buffer_type() == BufferType::DRAM;
+                tt::tt_metal::detail::AllocateBuffer(shared_buffer_ptr.get(), bottom_up);
+                shared_buffer_ptr->allocated_ = true; // Notify address getter that this buffer was allocated
                 // The address inserted here, will be used during asynchronous deallocate
-                GLOBAL_BUFFER_ADDRESS_MAP.insert(this->buffer_id, this->address());
-
+                Buffer::buf_id_to_address_map.insert({shared_buffer_ptr->buffer_instance_id, shared_buffer_ptr->address()});
             });
         }
+
         void deallocate() {
             if (this->device() == nullptr or not this->device()->initialized_ or this->size() == 0) {
                 return;
             }
             this->set_size(0);
             TT_ASSERT(this->device()->allocator_ != nullptr, "Expected allocator to be initialized!");
             // Extract the required buffer attributes from main thread (these are guaranteed to be correctly populated) and send to worker
-            this->device()->push_work([dev = this->device(), id = this->buffer_id, type = this->buffer_type()] () mutable {
-                // At this point, the address for this buffer has made it to GLOBAL_BUFFER_ADDRESS_MAP, since the worker has allocated the buffer.
-                tt::tt_metal::allocator::deallocate_buffer(*(dev->allocator_), GLOBAL_BUFFER_ADDRESS_MAP.buffer_address(id), type);
-                GLOBAL_BUFFER_ADDRESS_MAP.erase(id);
+            this->device()->push_work([dev = this->device(), id = this->buffer_instance_id, type = this->buffer_type()] () mutable {
+                // At this point, the address for this buffer has made it to buf_id_to_address_map, since the worker has allocated the buffer.
+                tt::tt_metal::allocator::deallocate_buffer(*(dev->allocator_), Buffer::buf_id_to_address_map.at(id), type);
+                Buffer::buf_id_to_address_map.erase(id);
             });
         }
+    private:
+        // Thread local variable used to extract buffer address from its id (used for asynchronous deallocation)
+        inline static thread_local std::unordered_map<uint64_t, uint32_t> buf_id_to_address_map = {};
```

Comment:
we need to stop using global variables like this in production unless it is for debug/tracing/profiling/optimizing purposes.
Making them thread_local

Reply:
`thread_local` allows each thread to track its own buffer addresses in a lock free manner, and independently resolve them in the destructor. `static` is used to allow destructors of buffer instances to look up the address based on the buffer id, without adding such a data-structure to the `tt_metal::Device` class, since that breaks abstraction.

I'm open to suggestions that allow us to implement lock free look ups, with clean abstractions.

Reply:
Let me think a little bit more about it. I think I have an idea :)